src/target/riscv/riscv-013.c

```diff
@@ -387,52 +387,47 @@ static unsigned int decode_dm(char *text, unsigned int address, unsigned int dat
 	return 0;
 }
 
-static unsigned int decode_dmi(struct target *target, char *text, unsigned int address,
-		unsigned int data)
-{
-	dm013_info_t *dm = get_dm(target);
-	if (!dm) {
-		if (text)
-			text[0] = '\0';
-		return 0;
-	}
-	return decode_dm(text, address - dm->base, data);
-}
-
-static void dump_field(struct target *target, int idle, const struct scan_field *field, bool discard_in)
+void riscv_dump_field(const struct target *target, int idle, const struct scan_field *field, bool discard_in)
 {
 	static const char * const op_string[] = {"-", "r", "w", "?"};
 	static const char * const status_string[] = {"+", "?", "F", "b"};
 
 	if (debug_level < LOG_LVL_DEBUG)
 		return;
 
-	uint64_t out = buf_get_u64(field->out_value, 0, field->num_bits);
-	unsigned int out_op = get_field(out, DTM_DMI_OP);
-	unsigned int out_data = get_field(out, DTM_DMI_DATA);
-	unsigned int out_address = out >> DTM_DMI_ADDRESS_OFFSET;
-
-	uint64_t in = buf_get_u64(field->in_value, 0, field->num_bits);
-	unsigned int in_op = get_field(in, DTM_DMI_OP);
-	unsigned int in_data = get_field(in, DTM_DMI_DATA);
-	unsigned int in_address = in >> DTM_DMI_ADDRESS_OFFSET;
-
-	log_printf_lf(LOG_LVL_DEBUG, __FILE__, __LINE__, __func__,
-			"%db %s %08x @%02x -> %s %08x @%02x; %di",
-			field->num_bits, op_string[out_op], out_data, out_address,
-			status_string[in_op], in_data, in_address, idle);
-
+	assert(field->out_value);
+	const uint64_t out = buf_get_u64(field->out_value, 0, field->num_bits);
+	const unsigned int out_op = get_field(out, DTM_DMI_OP);
+	const unsigned int out_data = get_field(out, DTM_DMI_DATA);
+	const unsigned int out_address = out >> DTM_DMI_ADDRESS_OFFSET;
+
+	if (field->in_value) {
+		const uint64_t in = buf_get_u64(field->in_value, 0, field->num_bits);
+		const unsigned int in_op = get_field(in, DTM_DMI_OP);
+		const unsigned int in_data = get_field(in, DTM_DMI_DATA);
+		const unsigned int in_address = in >> DTM_DMI_ADDRESS_OFFSET;
+
+		LOG_DEBUG("%db %s %08x @%02x -> %s %08x @%02x; %di",
+				field->num_bits, op_string[out_op], out_data, out_address,
+				status_string[in_op], in_data, in_address, idle);
+
+		if (!discard_in && in_op == DTM_DMI_OP_SUCCESS) {
+			char in_decoded[decode_dmi(target, NULL, in_address, in_data) + 1];
+			decode_dmi(target, in_decoded, in_address, in_data);
+			/* TODO: The decoded value is based on the assumtion
+			 * that DMI address was not changed between operations.
+			 * This is true for Spike, but may differ on HW.*/
```

Comment:
Can you please check if I understood the comment correctly.

If that is correct understanding, I would recommend to extend the comment, something along these lines:

```suggestion
			/* FIXME: OpenOCD should keep track of the last DMI address from which
			 * we have initiated a DMI read so that the obtained data can always be
			 * correctly decoded.
			 * The current code assumes that the hardware provides the read address 
			 * in the dmi.address field when returning the dmi.data. That is however not
			 * required by the spec, and therefore not guaranteed. Spike ISA simulator 
			 * behaves like that but other RISC-V debug implementations may not. 
			 */
```

Alternatively, an issue can be opened with the verbose description, and a only a short FIXME comment + the issue link can be put into the code.

Reply:
Addressed.